packages/wrangler/src/publish.ts

```diff
@@ -164,14 +157,45 @@ export default async function publish(props: Props): Promise<void> {
     let migrations;
     if (config.migrations.length > 0) {
       // get current migration tag
-      const scripts = await fetchResult<
-        { id: string; migration_tag: string }[]
-      >(`/accounts/${accountId}/workers/scripts`);
-      const script = scripts.find(({ id }) => id === scriptName);
+      type ScriptData = { id: string; migration_tag?: string };
+      let script: ScriptData | undefined;
+      if (!props.legacyEnv) {
+        try {
+          if (props.env) {
+            const scriptData = await fetchResult<{
+              script: ScriptData;
+            }>(
+              `/accounts/${accountId}/workers/services/${scriptName}/environments/${props.env}`
+            );
+            script = scriptData.script;
+          } else {
+            const scriptData = await fetchResult<{
+              default_environment: {
+                script: ScriptData;
+              };
+            }>(`/accounts/${accountId}/workers/services/${scriptName}`);
+            script = scriptData.default_environment.script;
+          }
+        } catch (err) {
+          if (
+            [10090, 10092].includes((err as { code: number }).code) === false
+          ) {
+            throw err;
+          }
+          // else it's a 404, no script found, and we can proceed
+        }
```

Comment:
How does this work if someone publishes using legacy environments, and then updates to use service environments? Will their old migrations be lost?

Reply:
In general, we have a pretty sucky story for migrating between wrangler <-> service environments (which is to say, we don't have anything at all right now). As you can imagine, there's also a data loss story here, since even if we did migrations correctly (which we should, assuming folks never delete previous migrations) the data from the legacy environment doesn't get carried over. A similar problem happens with secrets, since they don't move over either. This is part of a broader conversation we're having about service environments not being fully fleshed out, and hence the scary warning when you try to use them right now. 

That shouldn't block this specific PR tho.